### PR TITLE
Add Generalized Envy Graph algorithm

### LINKED ISSUE #141: Generalized Envy Graph

### PR DESCRIPTION
**Pull request by ketphan02**
Added new algorithm with the new algo configs

Closes #141

### CODE REVIEW

File: api/ai/new/geg_algorithm/geg_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
+import math
```
Comment (ketphan02) [inline api/ai/new/geg_algorithm/geg_algorithm.py:1]: Add algorithm pseudocode at top

File: api/ai/new/geg_algorithm/geg_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+            if i_star == -1:
+                raise ValueError("No i_star found")
+
+            self.allocation[i_star].append(student.id)
+            self.envy_graph.update_envy_graph(
+                i_star, self.utilities.get((i_star, student.id))
+            )
+
+            # While G(pi) contains directed cycle C do
+            # allocation_C = pi(i) if i not in C else pi(i_j+1) if i == i_j in C
```
Comment (ketphan02) [inline api/ai/new/geg_algorithm/geg_algorithm.py:270]:
```suggestion
```

File: api/ai/new/geg_algorithm/geg_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+                                i, self.utilities.get((i, student.id))
+                            )
+                        else:
+                            i_j = cycle.index(i)
+                            allocation_C.append(self.allocation[i_j + 1])
+                            self.envy_graph.update_envy_graph(
+                                i_j + 1, self.utilities.get((i_j + 1, student.id))
+                            )
+
+                    # Find the student with the lowest utility in the cycle
```
Comment (ketphan02) [inline api/ai/new/geg_algorithm/geg_algorithm.py:288]:
```suggestion
```

File: api/ai/new/geg_algorithm/geg_algorithm.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+
+        G = nx.DiGraph()
+        G.add_nodes_from(self.graph.keys())
+        for source, targets in self.graph.items():
+            for target in targets:
+                G.add_edge(source, target)
+
+        nx.draw(G, with_labels=True)
+        plt.show()
+
```
Comment (ketphan02) [inline api/ai/new/geg_algorithm/geg_algorithm.py:163]:
```suggestion
```

File: api/ai/new/geg_algorithm/geg_algorithm.py [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+        )
+        team_set = geg_algorithm.generate(student_provider.get())
+
+        # Print team set
+        for team in team_set.teams:
+            team_str = ""
+            for people in team.students:
+                team_str += f"{people.id} "
+            print(team_str)
+            print("-----")
```
Comment (ketphan02) [inline api/ai/new/geg_algorithm/geg_algorithm.py:369]: Use the new save method

**Comment by ketphan02** [comment]
Add documentation of algorithm to Latex

File: api/ai/new/geg_algorithm/geg_algorithm.py [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+                    )
+                else:
+                    student_utilities -= 1
+
+            utilities[(project.id, student.id)] = student_utilities
+
+    return utilities
+
+
+class EnvyGraph:
```
Comment (Opeyem1a) [inline api/ai/new/geg_algorithm/geg_algorithm.py:53]: each algorithm gets its own directory so you really don't have to crunch every class and util function into this one file, it feels kinda messy to me

File: api/ai/new/geg_algorithm/geg_algorithm.py [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+                            i_j = cycle.index(i)
+                            allocation_C.append(self.allocation[i_j + 1])
+                            self.envy_graph.update_envy_graph(
+                                i_j + 1, self.utilities.get((i_j + 1, student.id))
+                            )
+
+        return self.construct_team_from_allocation()
+
+
+if __name__ == "__main__":
```
Comment (Opeyem1a) [inline api/ai/new/geg_algorithm/geg_algorithm.py:276]: same comment as with MRR - why is this here?

File: api/ai/new/geg_algorithm/geg_algorithm.py [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+    ) -> Dict[Tuple[int, int], int]:
+        utilities: Dict[Tuple[int, int], int] = {}
+
+        for project in self.projects:
+            for student in students:
+                utilities[(project.id, student.id)] = random.randint(-5, 5)
+
+        return utilities
+
+    def _get_N_plus(self, projects: List[Project], student: Student) -> List[Project]:
```
Comment (Opeyem1a) [inline api/ai/new/geg_algorithm/geg_algorithm.py:199]: I'm guessing you named these functions/variables in a way that aligns with the formulas in the papers you read, but remember this is code that someone else will read, and some of these function/variable names make kinda no sense.

I see a few options:
1. Keep wacky function names, add better comments
2. Make better function names, add comments to explain what they correlate to in the papers
3. ??? be creative

File: api/ai/new/geg_algorithm/geg_algorithm.py [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+
+        This run in O(N)
+        """
+        vertices = [student.id for student in students]
+        self.graph: Dict[int, Set[int]] = {vertex: set([]) for vertex in vertices}
+        self.additive_utilities: Dict[int, int] = {vertex: 0 for vertex in vertices}
+
+    def add_edge(self, envy_project: int, other_project: int) -> None:
+        """
+        This run in O(1)
```
Comment (Opeyem1a) [inline api/ai/new/geg_algorithm/geg_algorithm.py:66]: nit;

Do we really need to annotate every function with its big-O? What value do these add as comments?

File: api/ai/new/interfaces/algorithm_options.py [hunk 10]
```diff
@@ -72,3 +72,9 @@ def validate(self):
 class PathGaspAlgorithmOptions(AlgorithmOptions):
     def validate(self):
         super().validate()
+
+
+@dataclass
+class GeneralizedEnvyGraphAlgorithmOptions(AlgorithmOptions):
```
Comment (Opeyem1a) [inline api/ai/new/interfaces/algorithm_options.py:78]: new algorithms should be added to the AlgorithmType enum

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

File: api/ai/new/geg_algorithm/geg_algorithm.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+                            self.envy_graph.update_envy_graph(
+                                i, self.utilities.get((i, student.id))
+                            )
+                        else:
+                            i_j = cycle.index(i)
+                            allocation_C.append(self.allocation[i_j + 1])
+                            self.envy_graph.update_envy_graph(
+                                i_j + 1, self.utilities.get((i_j + 1, student.id))
+                            )
+
```
Comment (ketphan02) [inline api/ai/new/geg_algorithm/geg_algorithm.py:159]: Algo not completed

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

File: api/ai/new/geg_algorithm/utils.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+            1 if student.meets_requirement(requirement) else -1
+            for requirement in requirements
+        ]
+    )
+
+
+def get_additive_utilities(
+    students: List[Student], projects: List[Project]
+) -> Dict[Tuple[int, int], int]:
+    utilities: Dict[Tuple[int, int], int] = {}
```
[the remark this answers is not in the document]
Comment (ketphan02) [reply api/ai/new/geg_algorithm/utils.py:20]: I agree, I was doing this so I could remove the check if project id exists, but it turns nasty real quick

**Review by Opeyem1a — APPROVED** [review]
LGTM

**Comment by Opeyem1a** [comment]
Current dependencies on/for this PR:
* main
  * **PR #140** <a href="https://app.graphite.dev/github/pr/Teamable-Analytics/algorithms/140?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/Teamable-Analytics/algorithms/140?utm_source=stack-comment).

**Pull request merged**